[FSDP] Make recursive_wrap, wrap APIs independent of ConfigAutoWrap.

rohan-varma opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #69358
* #69357
* #69356
* __->__ #68776

Makes these APIs independent of ConfigAutoWrap so that they can be
used by FSDP ctor without it knowing about ConfigAutoWrap. Although this has been done in a way such that the APIs are fully backward compatible.

Also gets us one step closer to killing ConfigAutoWrap.recursive_wrap and
auto_wrap(), as we will only support enable_wrap() and wrap() moving forward.

ConfigAutoWrap.recursive_wrap and wrap() are refactored to reduce code duplication. 

Will test via unittests and FSDP benchmarks to ensure the wrapping still works.

Closes https://github.com/pytorch/pytorch/issues/68241

Differential Revision: [D32604021](https://our.internmc.facebook.com/intern/diff/D32604021/)

cc @pietern @mrshenli @pritamdamania87 @zhaojuanmao @satgera @rohan-varma @gqchen @aazzolini @osalpekar @jiayisuse @SciPioneer @H-Huang

rohan-varma committed:
[FSDP] Make recursive_wrap, wrap APIs independent of ConfigAutoWrap.

Makes these APIs independent of ConfigAutoWrap so that they can be
used by FSDP ctor without it knowing about ConfigAutoWrap.

Also gets us one step closer to killing ConfigAutoWrap.recursive_wrap and
auto_wrap(), as we will only support enable_wrap() and wrap() moving forward.

Will test via unittests and FSDP benchmarks to ensure the wrapping still works.

Differential Revision: [D32604021](https://our.internmc.facebook.com/intern/diff/D32604021/)

[ghstack-poisoned]

rohan-varma committed:
Update on "[FSDP] Make recursive_wrap, wrap APIs independent of ConfigAutoWrap."


Makes these APIs independent of ConfigAutoWrap so that they can be
used by FSDP ctor without it knowing about ConfigAutoWrap. Although this has been done in a way such that the APIs are fully backward compatible.

Also gets us one step closer to killing ConfigAutoWrap.recursive_wrap and
auto_wrap(), as we will only support enable_wrap() and wrap() moving forward.

ConfigAutoWrap.recursive_wrap and wrap() are refactored to reduce code duplication. 

Will test via unittests and FSDP benchmarks to ensure the wrapping still works.

Differential Revision: [D32604021](https://our.internmc.facebook.com/intern/diff/D32604021/)

cc pietern mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse @SciPioneer @H-Huang

[ghstack-poisoned]

rohan-varma committed:
Update on "[FSDP] Make recursive_wrap, wrap APIs independent of ConfigAutoWrap."


Makes these APIs independent of ConfigAutoWrap so that they can be
used by FSDP ctor without it knowing about ConfigAutoWrap. Although this has been done in a way such that the APIs are fully backward compatible.

Also gets us one step closer to killing ConfigAutoWrap.recursive_wrap and
auto_wrap(), as we will only support enable_wrap() and wrap() moving forward.

ConfigAutoWrap.recursive_wrap and wrap() are refactored to reduce code duplication. 

Will test via unittests and FSDP benchmarks to ensure the wrapping still works.

Closes https://github.com/pytorch/pytorch/issues/68241

Differential Revision: [D32604021](https://our.internmc.facebook.com/intern/diff/D32604021/)

cc pietern mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse @SciPioneer @H-Huang

[ghstack-poisoned]

rohan-varma committed:
Update on "[FSDP] Make recursive_wrap, wrap APIs independent of ConfigAutoWrap."


Makes these APIs independent of ConfigAutoWrap so that they can be
used by FSDP ctor without it knowing about ConfigAutoWrap. Although this has been done in a way such that the APIs are fully backward compatible.

Also gets us one step closer to killing ConfigAutoWrap.recursive_wrap and
auto_wrap(), as we will only support enable_wrap() and wrap() moving forward.

ConfigAutoWrap.recursive_wrap and wrap() are refactored to reduce code duplication. 

Will test via unittests and FSDP benchmarks to ensure the wrapping still works.

Closes https://github.com/pytorch/pytorch/issues/68241

Differential Revision: [D32604021](https://our.internmc.facebook.com/intern/diff/D32604021/)

cc pietern mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse @SciPioneer @H-Huang

[ghstack-poisoned]

zhaojuanmao commented on torch/distributed/_fsdp/wrap.py:
```diff
@@ -161,16 +161,89 @@ def wrap(module: nn.Module, **wrap_overrides: Any) -> nn.Module:
+    return wrapper_cls(module, **wrap_overrides)
+
+
+def recursive_wrap(
```
nit: make it as private as well?

zhaojuanmao commented on torch/distributed/_fsdp/wrap.py:
```diff
@@ -161,16 +161,89 @@ def wrap(module: nn.Module, **wrap_overrides: Any) -> nn.Module:
+    return module, 0
+
+
 def auto_wrap(module: nn.Module, auto_wrap_policy: Optional[Callable] = None, **kwargs: Any) -> nn.Module:
```
since no one is using this, we can just kill it?

rohan-varma replied:
I think right now our internal benchmarks are using it, I was going to remove it from those and then remove these APIs

rohan-varma replied:
benchmarks are not actually using it so i removed it in the above diffs

zhaojuanmao commented on torch/distributed/_fsdp/wrap.py:
```diff
@@ -257,7 +330,6 @@ def __exit__(self, exc_type: Any, exc_val: Any, exc_tb: Any) -> None:
     def recursive_wrap(
```
kill it since no one is using it

rohan-varma committed:
Update on "[FSDP] Make recursive_wrap, wrap APIs independent of ConfigAutoWrap."


Makes these APIs independent of ConfigAutoWrap so that they can be
used by FSDP ctor without it knowing about ConfigAutoWrap. Although this has been done in a way such that the APIs are fully backward compatible.

Also gets us one step closer to killing ConfigAutoWrap.recursive_wrap and
auto_wrap(), as we will only support enable_wrap() and wrap() moving forward.

ConfigAutoWrap.recursive_wrap and wrap() are refactored to reduce code duplication. 

Will test via unittests and FSDP benchmarks to ensure the wrapping still works.

Closes https://github.com/pytorch/pytorch/issues/68241

Differential Revision: [D32604021](https://our.internmc.facebook.com/intern/diff/D32604021/)

cc pietern mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse @SciPioneer @H-Huang

[ghstack-poisoned]

rohan-varma committed:
Update on "[FSDP] Make recursive_wrap, wrap APIs independent of ConfigAutoWrap."


Makes these APIs independent of ConfigAutoWrap so that they can be
used by FSDP ctor without it knowing about ConfigAutoWrap. Although this has been done in a way such that the APIs are fully backward compatible.

Also gets us one step closer to killing ConfigAutoWrap.recursive_wrap and
auto_wrap(), as we will only support enable_wrap() and wrap() moving forward.

ConfigAutoWrap.recursive_wrap and wrap() are refactored to reduce code duplication. 

Will test via unittests and FSDP benchmarks to ensure the wrapping still works.

Closes https://github.com/pytorch/pytorch/issues/68241

Differential Revision: [D32604021](https://our.internmc.facebook.com/intern/diff/D32604021/)

cc pietern mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse @SciPioneer @H-Huang

[ghstack-poisoned]

rohan-varma commented:
@zhaojuanmao Made the recursive_wrap private, planning to kill auto_wrap, ConfigAutoWrap.recursive_wrap in later diffs after removing them from our internal benchmarks.

rohan-varma committed:
Update on "[FSDP] Make recursive_wrap, wrap APIs independent of ConfigAutoWrap."


Makes these APIs independent of ConfigAutoWrap so that they can be
used by FSDP ctor without it knowing about ConfigAutoWrap. Although this has been done in a way such that the APIs are fully backward compatible.

Also gets us one step closer to killing ConfigAutoWrap.recursive_wrap and
auto_wrap(), as we will only support enable_wrap() and wrap() moving forward.

ConfigAutoWrap.recursive_wrap and wrap() are refactored to reduce code duplication. 

Will test via unittests and FSDP benchmarks to ensure the wrapping still works.

Closes https://github.com/pytorch/pytorch/issues/68241

Differential Revision: [D32604021](https://our.internmc.facebook.com/intern/diff/D32604021/)

cc pietern mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse @SciPioneer @H-Huang

[ghstack-poisoned]

rohan-varma committed:
Update on "[FSDP] Make recursive_wrap, wrap APIs independent of ConfigAutoWrap."


Makes these APIs independent of ConfigAutoWrap so that they can be
used by FSDP ctor without it knowing about ConfigAutoWrap. Although this has been done in a way such that the APIs are fully backward compatible.

Also gets us one step closer to killing ConfigAutoWrap.recursive_wrap and
auto_wrap(), as we will only support enable_wrap() and wrap() moving forward.

ConfigAutoWrap.recursive_wrap and wrap() are refactored to reduce code duplication. 

Will test via unittests and FSDP benchmarks to ensure the wrapping still works.

Closes https://github.com/pytorch/pytorch/issues/68241

Differential Revision: [D32604021](https://our.internmc.facebook.com/intern/diff/D32604021/)

cc pietern mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse @SciPioneer @H-Huang

[ghstack-poisoned]

rohan-varma committed:
Update on "[FSDP] Make recursive_wrap, wrap APIs independent of ConfigAutoWrap."


Makes these APIs independent of ConfigAutoWrap so that they can be
used by FSDP ctor without it knowing about ConfigAutoWrap. Although this has been done in a way such that the APIs are fully backward compatible.

Also gets us one step closer to killing ConfigAutoWrap.recursive_wrap and
auto_wrap(), as we will only support enable_wrap() and wrap() moving forward.

ConfigAutoWrap.recursive_wrap and wrap() are refactored to reduce code duplication. 

Will test via unittests and FSDP benchmarks to ensure the wrapping still works.

Closes https://github.com/pytorch/pytorch/issues/68241

Differential Revision: [D32604021](https://our.internmc.facebook.com/intern/diff/D32604021/)

cc pietern mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse @SciPioneer @H-Huang

[ghstack-poisoned]
